### PR TITLE
fix: preserve no-newline-at-EOF markers through parse and patch build

### LINKED ISSUE #9: Preserve no-newline-at-EOF markers through parse and patch build
## What to build

Make stage, unstage, and discard work correctly on files whose last line has no
trailing newline. Today every `\ No newline at end of file` marker is dropped
while parsing the diff, so the patch git-hunk rebuilds claims the file ends in a
newline. Two distinct failures result:

1. Editing the last line of a file that has no trailing newline fails outright
   with `patch does not apply`.
2. Changing a file from "ends with newline" to "no trailing newline" reports
   success but silently stages nothing (the staged index is unchanged).

The fix is to preserve the no-newline markers end-to-end (parse → hunk diff →
built patch) and treat them as non-change, non-context lines so they don't
corrupt addition/deletion counts, hunk-header line counts, or split offsets.
Do not append a trailing newline after a hunk that ends in a no-newline marker.

## Acceptance criteria

- [ ] `git-hunk stage <id>` succeeds when the hunk touches the last line of a
      file with no trailing newline, and the staged content matches the working
      tree byte-for-byte (no spurious `\n` added).
- [ ] Staging a newline→no-newline transition actually removes the trailing
      newline from the index (verify with `git show :file | xxd`).
- [ ] unstage and discard round-trip correctly for the same files.
- [ ] Addition/deletion counts in `list` are unchanged by the marker lines.
- [ ] e2e tests cover: edit-last-line-no-newline, newline→no-newline,
      no-newline→newline.

Repro (currently fails):

    printf 'a\nb\nc' > f.txt   # no trailing newline
    git add f.txt && git commit -m init
    printf 'a\nb\ncX' > f.txt
    git-hunk stage <id>        # -> error: patch does not apply

## Blocked by

None - can start immediately.

### PR DESCRIPTION
**Pull request by wkentaro**
Fixes #9.

Every `\ No newline at end of file` marker was dropped while parsing the diff, so the patch git-hunk rebuilt always claimed the file ended in a newline. Staging the last line of a no-newline file failed with `patch does not apply`, and a newline -> no-newline change reported success while silently staging nothing.

This preserves the markers end-to-end (parse -> hunk split -> built patch) and treats them as non-change, non-context lines so they never corrupt addition/deletion counts, hunk-header line counts, or split offsets. `show` line numbers and `-l` line selection skip markers consistently, and a marker is dropped only when its preceding line is.

A representational follow-up (model the marker as a property of its line rather than a free-floating body line) is tracked in #28.

## Test plan
- [x] e2e covers edit-last-line-no-newline, newline -> no-newline, no-newline -> newline, plus unstage/discard round-trips and `-l` selection: `tests/e2e/no_newline_test.py`
- [x] unit covers marker preservation, counts, and split offsets: `tests/unit/_hunk/no_newline_test.py`
- [x] `make lint` and `uv run pytest` (105 passed)

### CODE REVIEW

**Comment by wkentaro** [comment]
> *This was generated by AI during PR processing.*

**Verdict: recommend-merge** (high confidence)

The bug was `parse_diff` stripping every `\ No newline at end of file` line, so `build_patch` emitted patches that lied to `git apply` about trailing-newline state. The fix preserves markers in `hunk.diff`, has `_advance_offsets` and the context counter skip them so `@@` line numbers are not inflated, and `_filter_body_lines` ties each marker to its preceding line via `prev_kept` so `-l` selection neither drops a needed marker nor keeps an orphan. All three edge cases (old-only, new-only, both) are covered; the unit test asserting the exact `@@ -8,4 +8,4 @@` header would fail against the pre-fix code. No blocking issues.
